Boundary velocity residual postprocessor

alarshi opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

Plugin to compare the modeled velocity (strains) with the input velocity (strains) at the surface.

### Before your first pull request:

* [X] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [X] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [X] I have tested my new feature locally to ensure it is correct.

gassmoeller commented on source/postprocess/visualization/velocity_residual.cc:
```diff
@@ -0,0 +1,169 @@
+		    if (use_spherical_unit_vectors)
+			{
+		      for (unsigned int d = 0; d < dim; ++d)
+		    	velocity[d] = Utilities::AsciiDataBoundary<dim>::get_data_component(
```
This would then be:
```suggestion
		    	velocity[d] = ascii_data_lookup->get_data_component(
```

gassmoeller commented on source/postprocess/visualization/velocity_residual.cc:
```diff
@@ -0,0 +1,169 @@
+    	{
+          prm.enter_subsection("Visualization");
+    	  {
+		    Utilities::AsciiDataBase<dim>::declare_parameters(prm,
+						  "$ASPECT_SOURCE_DIR/data/initial-temperature/adiabatic-boundary/",
+						  "adiabatic_boundary.txt", "Surface properties");
```
When you no longer derive from `AsciiDataBase` replace this line with the content of the the function you call here.

gassmoeller commented on source/postprocess/visualization/velocity_residual.cc:
```diff
@@ -0,0 +1,169 @@
+	    prm.enter_subsection("Postprocess");
+		  prm.enter_subsection("Visualization");
+		  {
+			Utilities::AsciiDataBase<dim>::parse_parameters(prm, "Surface properties");
```
Same here

gassmoeller commented on source/postprocess/visualization/velocity_residual.cc:
```diff
@@ -0,0 +1,169 @@
+		{
+		  for (unsigned int q=0; q<computed_quantities.size(); ++q)
+		  {
+			Tensor<1,dim> velocity;
```
This name is currently easy to confuse with the velocity computed by ASPECT, could you rename this variable? Maybe `reference_viscosity` would be good, but feel free to choose another name that makes clear this is the velocity read from an outside file that acts as the reference for our residual computation.

gassmoeller commented on source/postprocess/visualization/velocity_residual.cc:
```diff
@@ -0,0 +1,169 @@
+		    	velocity[d] = Utilities::AsciiDataBoundary<dim>::get_data_component(
+		    	  							this->get_geometry_model().translate_symbolic_boundary_name_to_id("top"),
+		    	  							input_data.evaluation_points[q], d);
+		      velocity = Utilities::Coordinates::spherical_to_cartesian_vector(velocity, input_data.evaluation_points[q]);
```
We talked about the remaining rotations between the velocities. This line is the most likely point where this could be introduced. You will want to check if `velocity` that you read from the data file has the same order of columns  as this function expects (v_phi, v_theta), and if latitude and longitude are defined the same way (spherical vs geographical coordinates).

alarshi replied:
Thank you for the suggestion! The problem was in fact with reading the vel_north (GPS data) as vel_theta and not -vel_theta.

gassmoeller commented on source/postprocess/visualization/velocity_residual.cc:
```diff
@@ -0,0 +1,169 @@
+/*
```
`VelocityResidual` is a bit misleading as a name, because the postprocessor only computes the residual for the surface. I would suggest renaming it to `BoundaryVelocityResidual`, this would also open the door for future extensions to other boundaries.

gassmoeller commented on source/postprocess/visualization/velocity_residual.cc:
```diff
@@ -0,0 +1,169 @@
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(VelocityResidual,
+                                                  "velocity residual",
+                                                  "A visualization for caculating the surface velocity to compare"
```
Please extend the documentation a bit, mentioning that this outputs the difference between the surface velocities and some reference velocity model.

gassmoeller commented on include/aspect/postprocess/visualization/boundary_velocity_residual.h:
```diff
@@ -0,0 +1,144 @@
+/*
+  Copyright (C) 2011 - 2017 by the authors of the ASPECT code.
```
```suggestion
  Copyright (C) 2020 by the authors of the ASPECT code.
```

gassmoeller commented on include/aspect/postprocess/visualization/boundary_velocity_residual.h:
```diff
@@ -0,0 +1,144 @@
+*/
+
+
+#ifndef _aspect_postprocess_visualization_velocity_residual_h
+#define _aspect_postprocess_visualization_velocity_residual_h
```
```suggestion
#ifndef _aspect_postprocess_visualization_boundary_velocity_residual_h
#define _aspect_postprocess_visualization_boundary_velocity_residual_h
```

gassmoeller commented on include/aspect/postprocess/visualization/boundary_velocity_residual.h:
```diff
@@ -0,0 +1,144 @@
+        private:
+          bool use_spherical_unit_vectors;
+
+          bool use_gps_data;
```
Please add documentation about this parameter, e.g. "Determines whether the model velocities are compared against ascii data files, or a gplates model." I would like to rename this into `use_ascii_data`, because the plugin could be used for anything (not just gps velocities).

gassmoeller commented on include/aspect/postprocess/visualization/boundary_velocity_residual.h:
```diff
@@ -0,0 +1,144 @@
+
+          /**
+           * Directory in which the data files are present.
+           */
+          std::string ascii_data_directory;
+
+          /**
+           * Filename of data file. The file names can contain the specifiers %s
+           * and/or %c (in this order), meaning the name of the boundary and the
+           * number of the data file time step.
+           */
+          std::string ascii_data_file_name;
+
+          /**
+           * Scale the data by a scalar factor. Can be used to transform the
+           * unit of the data (if they are not specified in SI units (m/s or
+           * m/yr depending on the "Use years in output instead of seconds"
+           * parameter).
+           */
+          double ascii_scale_factor;
+
+          /**
+           * Directory in which the data files are present.
+           */
+          std::string gplates_data_directory;
+
+          /**
+           * Filename of data file. The file names can contain the specifiers %s
+           * and/or %c (in this order), meaning the name of the boundary and the
+           * number of the data file time step.
+           */
+          std::string gplates_data_file_name;
+
+          /**
+           * Scale the data by a scalar factor. Can be used to transform the
+           * unit of the data (if they are not specified in SI units (m/s or
+           * m/yr depending on the "Use years in output instead of seconds"
+           * parameter).
+           */
+          double gplates_scale_factor;
```
Can we unify these parameters? I.e. could we just have `data_directory`, `data_file_name`, and `scale_factor`? Then the single parameter "Use ascii data" would determine if we load a gplates or an ascii data file.

gassmoeller commented on include/aspect/postprocess/visualization/boundary_velocity_residual.h:
```diff
@@ -0,0 +1,144 @@
+           */
+          double gplates_scale_factor;
+
+          /**
+           * Parsed user input of point1 and point2 in gplates
+           */
+          Tensor<1,2> pointone;
+          Tensor<1,2> pointtwo;
+
```
These two variables are only used inside the `initalize` function and not read in from file. As long as we do not use them anywhere else, please put them inside that function.

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,260 @@
+/*
+  Copyright (C) 2011 - 2019 by the authors of the ASPECT code.
```
```suggestion
  Copyright (C) 2020 by the authors of the ASPECT code.
```

bangerth replied:
Still relevant.

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,260 @@
+      void
+	  BoundaryVelocityResidual<dim>::initialize ()
+	  {
+      // The input gpaltes file with default values of pointone and pointtwo
```
plates

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,260 @@
+	  {
+      // The input gpaltes file with default values of pointone and pointtwo
+	    pointone[0] = 1.570796; pointone[1] = 0;
+	    pointtwo[0] = 1.570796; pointtwo[1] = 1.570796;
```
please put this as one instruction per line, we try to avoid multiple instructions per line (as a convention)

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,260 @@
+	    pointone[0] = 1.570796; pointone[1] = 0;
+	    pointtwo[0] = 1.570796; pointtwo[1] = 1.570796;
+
+	   gplates_lookup =  std_cxx14::make_unique<BoundaryVelocity::internal::GPlatesLookup<dim> > (
+			   pointone, pointtwo);
+
+	   gplates_lookup->load_file(gplates_data_directory + gplates_data_file_name, this->get_mpi_communicator());
+
+	 // The input ascii table contains one data column (velocity components) in addition to the coordinate columns.
+	   ascii_data_lookup = std_cxx14::make_unique<Utilities::AsciiDataLookup<dim> >(dim, ascii_scale_factor);
+	   ascii_data_lookup->load_file(ascii_data_directory + ascii_data_file_name, this->get_mpi_communicator());
```
could you decide this here as:

```suggestion
if (use_ascii_data == false)
{
	   gplates_lookup =  std_cxx14::make_unique<BoundaryVelocity::internal::GPlatesLookup<dim> > (
			   pointone, pointtwo);

	   gplates_lookup->load_file(gplates_data_directory + gplates_data_file_name, this->get_mpi_communicator());
}
else
{
	 // The input ascii table contains one data column (velocity components) in addition to the coordinate columns.
	   ascii_data_lookup = std_cxx14::make_unique<Utilities::AsciiDataLookup<dim> >(dim, ascii_scale_factor);
	   ascii_data_lookup->load_file(ascii_data_directory + ascii_data_file_name, this->get_mpi_communicator());
	   }
```

gassmoeller commented on include/aspect/postprocess/visualization/boundary_velocity_residual.h:
```diff
@@ -0,0 +1,144 @@
+		  parse_parameters (ParameterHandler &prm) override;
+
+        private:
+          bool use_spherical_unit_vectors;
```
Please add documentation that this determines whether the data files are in spherical (r,phi,theta) or cartesian (x,y,z) velocities.

bangerth replied:
This is still necessary.

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,260 @@
+		  prm.leave_subsection();
+	    prm.leave_subsection();
+	  }
+
+
```
Please remove these empty lines

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,260 @@
+												  "A visualization output object that generates output for the velocity"
+												  "residual at the top surface. The residual is computed at each point as "
+												  "the difference of the modeled velocities from the input velocities at"
+												  "the surface for each vector component. The user has an option to choose"
+												  "the input data as the GPS velocities in the ASCII data format with columns "
+												  "in the same format as described for the initial temperature or compositions"
+												  "or a velocity filed computed from the GPlates program. ")
```
I would write this like this:

```suggestion
												  "the surface for each vector component. The user has an option to choose"
												  "the input data as ascii data files (e.g. GPS velocities) with columns "
												  "in the same format as described for the 'ascii data' boundary velocity plugin "
												  "or a velocity field computed from the GPlates program 
												  "as described in the "gplates" boundary velocity plugin.")
```

bangerth replied:
You'll need to add spaces at the end of each line (before the `"`) to make sure there is a space when the individual strings/lines are all merged into one.

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,260 @@
+	  BoundaryVelocityResidual<dim>::initialize ()
+	  {
+      // The input gpaltes file with default values of pointone and pointtwo
+	    pointone[0] = 1.570796; pointone[1] = 0;
```
try `numbers::PI/2` instead of the hard-coded number here and in the line below

bangerth commented on include/aspect/postprocess/visualization/boundary_velocity_residual.h:
```diff
@@ -0,0 +1,144 @@
+       * and computes a variable that represents the strain rate at every
+       * point. The scalar strain rate is defined as $\sqrt{ (\varepsilon -
+       * \tfrac 13 \textrm{trace}\ \varepsilon \mathbf 1) : \varepsilon -
+       * \tfrac 13 \textrm{trace}\ \varepsilon \mathbf 1}$.
```
I don't think this accurately describes what the class is doing :-)

alarshi replied:
Thank you for picking that! I just pushed the changes which describe the correct documentation and attempts to address some other suggestions left by Rene from my last commit.

bangerth commented on include/aspect/postprocess/visualization/boundary_velocity_residual.h:
```diff
@@ -0,0 +1,144 @@
+           */
+          static
+		  void
+		  declare_parameters (ParameterHandler &prm);
```
Indentation is off here. You want to run `make indent`.

bangerth commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,260 @@
+												  "A visualization output object that generates output for the velocity"
+												  "residual at the top surface. The residual is computed at each point as "
+												  "the difference of the modeled velocities from the input velocities at"
+												  "the surface for each vector component. The user has an option to choose"
```
difference between this *and what*?

gassmoeller commented:
@alarshi is this ready for another review? We do not know if you are still working on a PR or if you are waiting for us unless you leave us a comment.

alarshi commented:
@gassmoeller: Yes, Rene, it is ready for review.

gassmoeller commented on include/aspect/postprocess/visualization/boundary_velocity_residual.h:
```diff
@@ -0,0 +1,134 @@
+       */
+      template <int dim>
+      class BoundaryVelocityResidual
+        : public DataPostprocessorVector<dim>,
```
Now that #3676 is merged, could you also derive this class from `public SurfaceOnlyVisualization<dim>`? This will then only compute the information on the actual surface of the domain, and store the result in a separate output file `solution_surface`. To make this work you will have to rebase this pull request to a recent version of ASPECT.

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,243 @@
+            ascii_data_lookup = std_cxx14::make_unique<Utilities::AsciiDataLookup<dim> >(dim, scale_factor);
+            ascii_data_lookup->load_file(data_directory + data_file_name, this->get_mpi_communicator());
+          }
+
```
please remove this empty line

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,243 @@
+            for (unsigned int q=0; q<computed_quantities.size(); ++q)
+              {
+                Tensor<1,dim> data_velocity;
+                if (use_spherical_unit_vectors && use_ascii_data)
+                  {
+                    const std::array<double,dim> cartesian_position = this->get_geometry_model().
+                                                                      cartesian_to_natural_coordinates(input_data.evaluation_points[q]);
+
+                    Point<dim> internal_position;
+
+                    for (unsigned int d = 0; d < dim; ++d)
+                      internal_position[d] = cartesian_position[d];
+
+                    for (unsigned int d = 0; d < dim; ++d)
+                      data_velocity[d] = ascii_data_lookup->get_data(internal_position, d);
+
+                    data_velocity = Utilities::Coordinates::spherical_to_cartesian_vector(data_velocity, input_data.evaluation_points[q]);
+
+                    for (unsigned int d = 0; d < dim; ++d)
+                      computed_quantities[q](d) = (data_velocity[d] - input_data.solution_values[q][d] * velocity_scaling_factor) ;
+                  }
+                else if (use_ascii_data && !use_spherical_unit_vectors)
+                  {
+                    for (unsigned int d = 0; d < dim; ++d)
+                      {
+                        data_velocity[d] = ascii_data_lookup->get_data(input_data.evaluation_points[q], d);
+                        computed_quantities[q](d) = (data_velocity[d] - input_data.solution_values[q][d] * velocity_scaling_factor) ;
+                      }
+                  }
+                else
+                  {
+                    data_velocity =  gplates_lookup->surface_velocity(input_data.evaluation_points[q]);
+                    for (unsigned int d = 0; d < dim; ++d)
+                      computed_quantities[q](d) = (data_velocity[d] - input_data.solution_values[q][d]) * velocity_scaling_factor;
+                  }
```
I would write this approximately the following way. This way you do not need 3 branches, but only ascii data or gplates. Also it seems that at the moment you only support either : (i) sperical models and spherical unit vectors for velocity (branch 1), or (ii) cartesian models and no spherical unit vectors (branch 2). But with my version you also support ascii data files for spherical models, but prescribing cartesian velocities in the data columns.

```suggestion
if (use_ascii_data == true)
{
  Point<dim> internal_position = input_data.evaluation_points[q];
  
  if (this->get_geometry_model().natural_coordinate_system() == Utilities::Coordinates::spherical)
  {
    const std::array<double,dim> spherical_position = this->get_geometry_model().
      cartesian_to_natural_coordinates(input_data.evaluation_points[q]);

  for (unsigned int d = 0; d < dim; ++d)
    internal_position[d] = spherical_position[d];
  }

  for (unsigned int d = 0; d < dim; ++d)
    data_velocity[d] = ascii_data_lookup->get_data(internal_position, d);

  if (use_spherical_unit_vectors == true)
    data_velocity = Utilities::Coordinates::spherical_to_cartesian_vector(data_velocity, input_data.evaluation_points[q]);
}
else
{
  data_velocity =  gplates_lookup->surface_velocity(input_data.evaluation_points[q]);
}
                  
for (unsigned int d = 0; d < dim; ++d)
  computed_quantities[q](d) = (data_velocity[d] - input_data.solution_values[q][d]) * velocity_scaling_factor;
```

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,243 @@
+        {
+          prm.enter_subsection("Visualization");
+          {
+            prm.enter_subsection("Surface properties");
```
This should probably be `Boundary velocity residual`

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,243 @@
+                                 "Specify velocity as r, phi, and theta components "
+                                 "instead of x, y, and z. Positive velocities point up, east, "
+                                 "and north (in 3D) or out and clockwise (in 2D). "
+                                 "This setting only makes sense for spherical geometries.");
```
Add: "GPlates data is always interpreted to be in east and north directions and is not affected by this parameter."

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,243 @@
+                                 "This setting only makes sense for spherical geometries.");
+              prm.declare_entry ("Use ascii data", "false",
+                                 Patterns::Bool (),
+                                 "Use ascii data (e.g., GPS) for computing residual velocities .");
```
```suggestion
                                 "Use ascii data files (e.g., GPS) for computing residual velocities instead of GPlates data.");
```

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,243 @@
+        prm.enter_subsection("Postprocess");
+        prm.enter_subsection("Visualization");
+        {
+          prm.enter_subsection("Surface properties");
```
"Boundary velocity residual"

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,243 @@
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(BoundaryVelocityResidual,
+                                                  "boundary velocity residual",
+                                                  "A visualization output object that generates output for the velocity "
+                                                  "residual at the top surface. The residual is computed at each point at the "
```
double `the`

gassmoeller commented on source/postprocess/visualization/boundary_velocity_residual.cc:
```diff
@@ -0,0 +1,243 @@
+                                                  "boundary velocity residual",
+                                                  "A visualization output object that generates output for the velocity "
+                                                  "residual at the top surface. The residual is computed at each point at the "
+                                                  " the surface as the difference between the modeled velocities and the input "
```
I would say `input data velocities`.

alarshi commented:
@gassmoeller Hi Rene, I just pushed the changes and the test case. Please give me your feedback whenever you could :)

gassmoeller commented on cookbooks/composition-fields.prm:
```diff
@@ -0,0 +1,119 @@
+#########################################################
```
Please remove this file from the PR with `git rm  cookbooks/composition-fields.prm`.

gassmoeller commented on tests/ascii_data_boundary_velocity_residual_shell.prm:
```diff
@@ -0,0 +1,97 @@
+##### simple test for ascii data
```
Update the comment.

gassmoeller commented on tests/ascii_data_boundary_velocity_residual_shell/original.prm:
```diff
@@ -0,0 +1,98 @@
+##### simple test for ascii data
```
This file is not helpful as test output, because it will always be the same as the test .prm file. The same is true for the parameters.* files. We usually want to add files into the test ouput directory that will show us if something fails with the functionality of this test, thus `screen-output` and `statistics` are often added. For this test you would want to also add one of the `surface-solution` files to make sure the output of the visualization postprocessor stays the same.

gassmoeller commented on tests/ascii_data_boundary_velocity_residual_shell.prm:
```diff
@@ -0,0 +1,97 @@
+subsection Postprocess
+  set List of postprocessors = velocity statistics, temperature statistics, heat flux statistics, visualization
+  subsection  Visualization
+    set List of output variables      = adiabat, material properties, gravity, heat flux map, strain rate, boundary velocity residual
```
please remove everything but `boundary velocity residual` to make the output files as small as possible. Also switch the output format to 'gnuplot'. The default .vtu output is compressed binary data, and that is hard to compare against reference results (as soon as one number changes only slightly the whole output will be different), but .gnuplot data is stored as ascii text.

alarshi commented:
@gassmoeller Hi Rene, thank you for the suggestions. I pushed the revised version in the latest commits.

alarshi commented:
@gassmoeller : Hello Rene, I removed the compositional_fields.prm from tracking. I also forgot to add the entry into doc/module/changes, I added it in this commit. Thanks!

gassmoeller commented:
/rebuild

alarshi commented:
@gassmoeller Hello Rene, the tester gave an error because the screen-output had absolute paths, and not relative of the ascii data files. 
```##3       #:6   <== /home/asaxena/opt/Aspect2.0/aspect/data/boundary-velocity/ascii-data/test/shell_3d_bottom.0.txt.

##3       #:6   ==> ASPECT_DIR/data/boundary-velocity/ascii-data/test/shell_3d_bottom.0.txt. ```

I changed the paths in the recent commit and also the end time of the model run from 1e3 to 0, because we don't really need the next time steps.

Another error which tester had was about the output directory from ```ascii_data_boundary_velocity_residual_shell``` to output-ascii_data_boundary_velocity_residual_shell , such as

``` <== 0 0.000000000000e+00 0.000000000000e+00 24 972 305 0 102 106 212 3.76333584e-01 1.11973264e+00 2.73000000e+02 1.48854480e+03 3.00000000e+03 4.45744333e-01 -2.90672074e+13 -6.21208071e+13 -1.12506468e+15 1.13429758e+15 7.83775571e-03 ascii_data_boundary_velocity_residual_shell/solution/solution-00000 ascii_data_boundary_velocity_residual_shell/solution_surface/solution_surface-00000 

##24      ==> 0 0.000000000000e+00 0.000000000000e+00 24 972 305 0 102 106 212 3.76333584e-01 1.11973264e+00 2.73000000e+02 1.48854480e+03 3.00000000e+03 4.45744333e-01 -2.90672074e+13 -6.21208071e+13 -1.12506468e+15 1.13429758e+15 7.83775571e-03 output-ascii_data_boundary_velocity_residual_shell/solution/solution-00000 output-ascii_data_boundary_velocity_residual_shell/solution_surface/solution_surface-00000 ```

The folder ```output-ascii_data_boundary_velocity_residual_shell``` was created during ```setup_tests```.  Could you please help me how to tell the tester to look into  ```ascii_data_boundary_velocity_residual_shell``` directory?

gassmoeller commented:
Hi Arushi, yes these things happen because the tester replaces the output directory and file paths to make the tests independent of the system. The easiest fix is to replace the output files with the output from the reference tester as described at the end of this manual section: http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests.

gassmoeller commented:
Very nice, this is ready to be merged. I know we discussed waiting for the statistics counterpart to this postprocessor, but looking at it now I would prefer to have this in by now to avoid potential conflicts with test results or changes on the main branch. Since it is a very specific postprocessor I would not expect problems if we have to change the input parameters when we create the statistics postprocessor (soon).
